Improve backports documentation (#15589)
decidim/decidim: docs/modules/develop/pages/backports.adoc

We only offer support for the last two mayor version.

For instance, if we are on the version `0.32.0.dev`, we make backports for `v0.31.x` and `v0.30.x`. We only backport fixes and not features, as we want to respect xref:develop:guide_conventions.adoc[Semantic Versioning].

The idea of a backport is to bring a given fix from the last development version to an older release.

The process for making backports is the following:

. Reproduce the error against develop

> . Develop a fix for the issue, starting from the develop

Reviewer: ```suggestion
. Reproduce the error against develop.
. Develop a fix for the issue, starting from the develop.
```

> . Open the Pull Request, adding the following Red labels: `type: fix`, `release: 0.xx`, in our example using `0.32.0.dev`, we would attach `type: fix`, `release: v0.31`.

Reviewer: ```suggestion
. Open the Pull Request, adding the following labels: `type: fix`, `release: 0.xx`, in our example using `0.32.0.dev`, we would attach `type: fix`, `release: v0.31`.
```

> . If the fix affects multiple versions, attach the corresponding release tag ( ex: `release: v0.30`)

Reviewer: ```suggestion
. If the fix affects multiple versions, attach the corresponding release tag (for instance, `release: v0.30`).
```

. Be patient and wait for a maintainer to review and merge

> . Once it is merged, an automated workflow will attempt to apply the patch

Reviewer: ```suggestion
. Be patient and wait for a maintainer to review and merge.
. Once it is merged, an automated workflow will attempt to apply the patch.
```

> . If you have an already merged Pull Request, you can add another `release: v0.xx` tag, and this will retrigger the backport workflow. So you do not need to decide when you open the PR which are the versions you want to backport to.

Reviewer: ```suggestion
. If you have an already merged Pull Request, you can add another `release: v0.xx` tag, and this will retrigger the backport workflow. You do not need to decide when you open the PR which are the versions you want to backport to.
```

== Maintainers only

> in order to ease up the backport process, you will need to install the `decidim-maintainers_toolbox` gem

Reviewer: ```suggestion
In order to ease up the backport process, you will need to install the `decidim-maintainers_toolbox` gem.
```

[source,bash]
----
gem install decidim-maintainers_toolbox
----

> This will provide you with some commands that you can use for day to day activities:

Reviewer: I was going to say that we should only have it in one place with the documentation but I see this is already the place xD 

<img width="1097" height="788" alt="image" src="https://github.com/user-attachments/assets/edd0d086-86ec-42a8-b2c2-a8c7a6841454" />

> Before starting to use `decidim-maintainers_toolbox` gem, you will need to make sure that you have available the `gh` command, provided by `GitHub`. Follow the https://github.com/cli/cli[project's page] for installation steps that are most suited for your Operating System.

Reviewer: ```suggestion
Before starting to use `decidim-maintainers_toolbox` gem, you will need to make sure that you have available the `gh` command, provided by `GitHub`. Follow the https://github.com/cli/cli[gh project's page] for installation steps that are most suited for your Operating System.
```

=== decidim-backports-checker

> The `decidim-backports-checker` is an utility that you can use to investigate which are the Pull Requests that needs to be backported.

Reviewer: As the output for this one is basically 100% the whole idea of the tool, I think it'd be nice to have a screenshot of the ouput (not a blocker - just a nice to have).

We already have the output below, but mind that this tool relies in color coding, and that's not something that can't be done with asciidoctor/antora (or I couldn't find it)

[source, bash]
----
Usage:
  decidim-backports-checker --github-token=GITHUB_TOKEN --last-version-number=LAST_VERSION_NUMBER

Options:
--github-token=GITHUB_TOKEN                # Required. Github Personal Access Token (PAT). It can be obtained from https://github.com/settings/tokens/new. You will need to create one with `public_repo` access.
Alternatively, you can use the `gh` CLI tool to authenticate with `gh auth token` (i.e. --github-token=$(gh auth token))
--last-version-number=LAST_VERSION_NUMBER  # Required. The version number of the last supported version that you want to do the backports to. It must have the format MAJOR.MINOR.
[--days-to-check-from=N]                   # How many days since the pull requests were merged we will check from.
# Default: 90

Backports checker. Shows the status of the pull requests opened in the last days

The output by default is for the terminal. It uses a color code to differentiate the status of the Pull Requests, following the colors of GitHub:
- Purple: closed with a merge
- Red: closed without being merge
- Green: opened without being merge
----

> Any maintainer that wants to see the pending Pull Requests that needs to be backported to a specific version, should type in the following command:

Reviewer: Which command?

In our previous example, we have the develop branch set to be on 0.32.0.dev, which means that our latest released minor version is 0.31. We are using the minor version regardless if we are at `0.31.0.rc2`, `0.31.0` or `0.31.1`.

[source, bash]
----
decidim-backports-checker --github-token=$(gh auth token) --last-version-number=0.31
----

Running the above command would output the following data:

> [source, bash]

Reviewer: Another argument why I prefer the screenshot approach here:

We can use real world examples with real titles, and PR ids that will facilitate the understanding.

Also we don't need to worry about the table format. This is what it looks like now:

<img width="1236" height="263" alt="image" src="https://github.com/user-attachments/assets/36d6f661-d712-46fc-8a4c-6693a3f1edf1" />

----
PRs: |===============================================================================================================================================================================================================================|
|   ID   |                                        Title                                        | Backport v0.31 | Backport v0.30 |
|--------|-------------------------------------------------------------------------------------|----------------|----------------|
| #XXXXX | The name of the PR with Id #XXXXX                                              |     #AAAA     |      #CCCCC      |
| #YYYYY | The name of the PR with Id #YYYYY                     |     #BBBB     |      None      |
----

In my example, `#XXXXX` and `#YYYYY` are the Pull Requests opened against the `develop` branch, while `#AAAAA`, `#BBBBB` and `#CCCCC` are the actual backports targeting a specific version

> If you want to address a PR only for a specific version, you open the Pull Request for that version, then you can add `no-backport` label for the Pull Request that you do not want to backport to oldest version.

Reviewer: ```suggestion
If you want to address a PR only for a specific version, you open the Pull Request for that version, then you can add `no-backport` label for the Pull Request that you do not want to backport to the oldest version.
```

  --pull-request-id=PULL_REQUEST_ID                                                                       # Required. The ID of the pull request that you want to make the backport from. It should have the "type: fix" label.
  [--exit-with-unstaged-changes], [--no-exit-with-unstaged-changes], [--skip-exit-with-unstaged-changes]  # Optional. Whether the script should exit with an error if there are unstaged changes in the current project.
                                                                                                          # Default: true
  [--with-console], [--no-with-console], [--skip-with-console]                                            # Optional. Disables the shell dropout
                                                                                                          # Default: true

Backport a pull request to another branch
----

An example of the command is:

[source, bash]
----
decidim-backporter --github-token=$(gh auth token) --pull-request-id=XXXXX --version-number=0.31
----

> In this particular example, XXXXX is the Pull Request merged against develop branch, while `--version-number` is the version where we want to backport.

Reviewer: ```suggestion
decidim-backporter --github-token=$(gh auth token) --pull-request-id=1234 --version-number=0.31
----

In this particular example, 1234 is the Pull Request merged against develop branch, while `--version-number` is the version where we want to backport.
```

We do not need to pass the full version, as this will apply the fix only for the unreleased versions, regardless if we are on `0.31.0.rc1`, `0.31.0` or `0.31.1`

The command will attempt to apply the patch, and if the patch fails with error, a console will be displayed, allowing you to change the files in your editor. After the conflict is fixed, you will need to write the following in your console:
[source, bash]
----
git add <any conflicting files that you fixed>
git cherry-pick --continue

> type logout / press CTRL + D

Reviewer: ```suggestion
# type logout or press CTRL + D
```

----

> The process will continue with creating the backport Pull Request.

Reviewer: ```suggestion
The process will continue creating the backport Pull Request.
```

> If you want to skip the backport creation for any reason, you just need to ensure that when your console drops, you do not have staged / modified files. Typing a `git cherry-pick --abort` will clean the backport process. After that you just need to exit the console with either `logout` either `CTRL+D`.

Reviewer: ```suggestion
If you want to skip the backport creation for any reason, you just need to ensure that when your console drops, you do not have staged / modified files. Typing a `git cherry-pick --abort` will clean the backport process. After that you just need to exit the console with either `logout` or `CTRL+D`.
```
